openmdao/core/explicitcomponent.py

```diff
@@ -224,22 +251,45 @@ def _apply_linear(self, vec_names, mode, scope_out=None, scope_in=None):
         """
         with Recording(self.pathname + '._apply_linear', self.iter_count, self):
             for vec_name in vec_names:
+                if vec_name not in self._rel_vec_names:
+                    continue
+
                 with self._matvec_context(vec_name, scope_out, scope_in, mode) as vecs:
                     d_inputs, d_outputs, d_residuals = vecs
 
                     # Jacobian and vectors are all scaled, unitless
                     with self.jacobian_context() as J:
                         J._apply(d_inputs, d_outputs, d_residuals, mode)
 
+                    # if we're not matrix free, we can skip the bottom of
+                    # this loop because compute_jacvec_product does nothing.
+                    if not self.matrix_free:
```

Comment:
technically this logic is slightly flawed. If we're matrix free, it would still try do do the first J_apply. even though thats not technically needed.

Reply:
If I modify the routine so that if we're matrix free it skips the J.apply, there are about 30 failed tests.

Reply:
Correction, I had mis-tabbed my 'continue' statement.  Now 10 tests fail.

Reply:
hrm.... ok, I rescind my statement then. Its ok as is